metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala

```diff
@@ -0,0 +1,339 @@
+              packageParts,
+              s"$otherParts${d.name.value}#",
+              Some(s"$otherParts${d.name.value}."),
+            )
+          case d: Defn.Given =>
+            (packageParts, s"$otherParts${d.name.value}#", None)
+          case _ => (packageParts, otherParts, alternative)
+        }
+
+      enclosedChild(tree)
+        .map(loop(_, packageParts1, otherParts1, alternative1))
+        .getOrElse {
+          (packageParts1, otherParts1, alternative1)
+        }
+    }
+
+    trees
+      .get(path)
+      .map { tree =>
+        val (packagePart, otherParts, alternative) = loop(tree)
+        ContextSymbols(packagePart, otherParts, alternative)
+      }
+      .getOrElse(ContextSymbols.empty)
+
+  }
+
+}
+
+case class ScalaDocLink(value: String) {
+  def toScalaMetaSymbols(
+      contextSymbols: => ContextSymbols
+  ): List[ScalaDocLinkSymbol] =
+    if (value.isEmpty()) List.empty
+    else {
+      val symbol = symbolWithFixedPackages
+      val optIndexOfDot =
+        ScalaDocLink.findIndicesOf(value, List('.')).headOption
+      def all = List(symbol) ++
+        contextSymbols.withThis(symbol) ++
+        contextSymbols.withPackage(symbol)
+      val withPrefixes: List[String] =
+        optIndexOfDot
+          .map { indexOfDot =>
+            symbol.splitAt(indexOfDot + 1) match {
+              case ("this/", rest) => contextSymbols.withThis(rest)
+              case ("package/", rest) => contextSymbols.withPackage(rest)
+              case _ if symbol.contains("/") => List(symbol)
+              case _ => all
+            }
+          }
+          .getOrElse(all)
+
+      val indexOfLParen =
+        ScalaDocLink.findIndicesOf(symbol, List('(', '[')).headOption
+
+      if (indexOfLParen.nonEmpty) {
+        val toDrop = symbol.length - indexOfLParen.get
+        withPrefixes.flatMap(sym => List(MethodSymbol(sym.dropRight(toDrop))))
+      } else {
+        symbol.last match {
+          case '#' | '.' | '/' => withPrefixes.map(StringSymbol(_))
+          case '$' => // forces link to refer to a value (an object, a value, a given)
+            withPrefixes.flatMap(sym =>
+              List(
+                StringSymbol(s"${sym.dropRight(1)}."),
+                MethodSymbol(s"${sym.dropRight(1)}"),
+              )
+            )
+          case '!' => // forces link to refer to a type (a class, a type alias, a type member)
+            withPrefixes.flatMap(sym =>
+              List(StringSymbol(s"${sym.dropRight(1)}#"))
+            )
+          case _ =>
+            withPrefixes.flatMap(sym =>
+              List(
+                StringSymbol(s"$sym#"),
+                StringSymbol(s"$sym."),
+                MethodSymbol(sym),
+              )
+            )
+        }
+      }
+    }
+
+  private def symbolWithFixedPackages = {
+    val fixedPackages =
+      ScalaDocLink
+        .splitAt(value, '.')
+        .map { str =>
+          if (
+            str.headOption.exists(_.isLower) ||
```

Comment:
Yes, and those will be a problem. I'm not sure how to go about this. If we don't assume that packages start with lower case, we'd have to produce quite a lot of possible symbols. Maybe not that much, but I could easily get to ~50 to produce for possible scopes and types (class, object, method). I guess we could try a different approach and instead of producing the symbols create a fuzzy symbol search in `GlobalSymbolIndex`.

Reply:
Let's get back to this as a follow back